File: torch-c-article/lib/th/TestIdentityModule.lua
require('torch')
require('nn')

--- @class TestIdentityModule
local TestIdentityModule, TestIdentityModuleParent = torch.class('nn.TestIdentityModule', 'nn.Module')

--- Initialize.
function TestIdentityModule:__init()

end

--- Compute forward pass.
-- @param input layer input
-- @param output
function TestIdentityModule:updateOutput(input)
  self.output = input:clone()

  if input:type() == 'torch.FloatTensor' then
    assert(lib.cpu)
    lib.cpu.test_identity_module_updateOutput(5, input:size():data(), input:data(), self.output:data())
  elseif input:type() == 'torch.CudaTensor' then
    assert(lib.gpu)
    lib.gpu.test_identity_module_updateOutput(5, input:size():data(), input:data(), self.output:data())
  else
    assert(false)
  end

  return self.output
end

--- Compute the backward pass.
-- @param input original input
-- @param gradOutput gradients of top layer
-- @return gradients with respect to input
function TestIdentityModule:updateGradInput(input, gradOutput)
  self.gradInput = gradOutput:clone()

  if input:type() == 'torch.FloatTensor' then
    assert(lib.cpu)
    lib.cpu.test_identity_module_updateGradInput(5, input:size():data(), input:data(), gradOutput:data(), self.gradInput:data())
  elseif input:type() == 'torch.CudaTensor' then
    assert(lib.gpu)
    lib.gpu.test_identity_module_updateGradInput(5, input:size():data(), input:data(), gradOutput:data(), self.gradInput:data())
  else
    assert(false)
  end

  return self.gradInput
end

File: torch-c-article/lib/cpp/cpu/test_identity_module.cpp
#include <cstdio>
#include <cassert>
#include "test_identity_module.h"

void test_identity_module_updateOutput(const int rank, const long* dims, const float* input, float* output) {
  assert(rank == 5);

  const int batch_size = dims[0];
  const int channels = dims[1];
  const int height = dims[2];
  const int width = dims[3];
  const int depth = dims[4];

  for (int b = 0; b < batch_size; b++) {
    for (int c = 0; c < channels; c++) {
      for (int h = 0; h < height; h++) {
        for (int w = 0; w < width; w++) {
          for (int d = 0; d < depth; d++) {
            output[(((b*channels + c)*height + h)*width + w)*depth + d] = input[(((b*channels + c)*height + h)*width + w)*depth + d];
          }
        }
      }
    }
  }
}

void test_identity_module_updateGradInput(const int rank, const long* dims, const float* input, const float* grad_output, float* grad_input) {
  assert(rank == 5);

  const int batch_size = dims[0];
  const int channels = dims[1];
  const int height = dims[2];
  const int width = dims[3];
  const int depth = dims[4];

  for (int b = 0; b < batch_size; b++) {
    for (int c = 0; c < channels; c++) {
      for (int h = 0; h < height; h++) {
        for (int w = 0; w < width; w++) {
          for (int d = 0; d < depth; d++) {
            grad_input[(((b*channels + c)*height + h)*width + w)*depth + d] = grad_output[(((b*channels + c)*height + h)*width + w)*depth + d];
          }
        }
      }
    }
  }
}